tests/fuzz/sequence_compression_api.c

```diff
@@ -0,0 +1,258 @@
+                            const void *src, size_t srcSize,
+                            const void *dict, size_t dictSize,
+                            const ZSTD_Sequence* generatedSequences, size_t generatedSequencesSize,
+                            size_t wLog, unsigned cLevel, unsigned hasDict)
+{
+    size_t cSize;
+    size_t dSize;
+    ZSTD_CDict* cdict = NULL;
+    ZSTD_DDict* ddict = NULL;
+
+    ZSTD_CCtx_reset(cctx, ZSTD_reset_session_and_parameters);
+    ZSTD_CCtx_setParameter(cctx, ZSTD_c_nbWorkers, 0);
+    ZSTD_CCtx_setParameter(cctx, ZSTD_c_compressionLevel, cLevel);
+    ZSTD_CCtx_setParameter(cctx, ZSTD_c_windowLog, wLog);
+    /* TODO: Add block delim mode fuzzing */
+    ZSTD_CCtx_setParameter(cctx, ZSTD_c_blockDelimiters, ZSTD_sf_noBlockDelimiters);
+    if (hasDict) {
+        cdict = ZSTD_createCDict(dict, dictSize, cLevel);
+        FUZZ_ASSERT(cdict);
+        ZSTD_CCtx_refCDict(cctx, cdict);
+
+        ddict = ZSTD_createDDict(dict, dictSize);
+        FUZZ_ASSERT(ddict);
+        ZSTD_DCtx_refDDict(dctx, ddict);
+    }
+
+    cSize = ZSTD_compressSequences(cctx, compressed, compressedCapacity,
+                                   generatedSequences, generatedSequencesSize,
+                                   src, srcSize);
+    FUZZ_ZASSERT(cSize);
+    dSize = ZSTD_decompressDCtx(dctx, result, resultCapacity, compressed, cSize);
+    FUZZ_ZASSERT(dSize);
+
+    if (cdict) {
+        ZSTD_freeCDict(cdict);
+    }
+    if (ddict) {
+        ZSTD_freeDDict(ddict);
+    }
+    return dSize;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *src, size_t size)
+{
+    void* rBuf;
+    size_t rBufSize;
+    void* cBuf;
+    size_t cBufSize;
+    void* generatedSrc;
+    size_t generatedSrcSize;
+    ZSTD_Sequence* generatedSequences;
+    size_t nbSequences;
+    void* literalsBuffer;
+    size_t literalsSize;
+    void* dictBuffer;
+    size_t dictSize = 0;
+    unsigned hasDict;
+    unsigned wLog;
+    int cLevel;
+
+    FUZZ_dataProducer_t *producer = FUZZ_dataProducer_create(src, size);
+    literalsSize = FUZZ_dataProducer_uint32Range(producer, 1, ZSTD_FUZZ_GENERATED_LITERALS_MAXSIZE);
+    literalsBuffer = FUZZ_malloc(literalsSize);
+    literalsBuffer = generatePseudoRandomString(literalsBuffer, literalsSize);
+
+    hasDict = FUZZ_dataProducer_uint32Range(producer, 0, 1);
+    if (hasDict) {
+        dictSize = FUZZ_dataProducer_uint32Range(producer, 1, ZSTD_FUZZ_GENERATED_DICT_MAXSIZE);
+        dictBuffer = FUZZ_malloc(dictSize);
+        dictBuffer = generatePseudoRandomString(dictBuffer, dictSize);
+    }
+    // Generate window log first so we dont generate offsets too large
+    wLog = FUZZ_dataProducer_uint32Range(producer, ZSTD_WINDOWLOG_MIN, ZSTD_WINDOWLOG_MAX);
+    cLevel = FUZZ_dataProducer_int32Range(producer, (int)ZSTD_minCLevel, (int)ZSTD_maxCLevel);
+
+    generatedSequences = FUZZ_malloc(sizeof(ZSTD_Sequence)*ZSTD_FUZZ_GENERATED_SRC_MAXSIZE);
```

Comment:
Woah this is a huge allocation. Like 512MB I think. Could we put a cap on the number of sequences, like 2^17?

Also, this allocation, and `generatedSrc` can be shared across multiple calls to `LLVMFuzzerTestOneInput()`.

Also, for all allocations you should `FUZZ_ASSERT(generatedSequences != NULL)`.

Reply:
Yikes you're right, that's a bit excessive.
Generally should we just make all of these buffers global to reduce allocations? They're all pretty big.

Is the additional `FUZZ_ASSERT()` necessary? It looks like `FUZZ_MALLOC()` has a `FUZZ_ASSERT()` in it, unless we're allocating 0 bytes.

Reply:
> Generally should we just make all of these buffers global to reduce allocations? 

Allocations of a fixed size can be global. If they are dynamically sized they should be local. E.g. the buffer that we decompress into should be local because we want it to be exactly sized, so ASAN can catch buffer over-runs.

> Is the additional FUZZ_ASSERT() necessary? It looks like FUZZ_MALLOC() has a FUZZ_ASSERT() in it, unless we're allocating 0 bytes.a

Oh good point, I forgot about that!